File: BOJ/1918.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int p(char a) { // priority
	if (a == '*' || a == '/') return 1;
	else if (a == '+' || a == '-') return 2;
	else if (a == ')') return 3;
	else if (a == '(') return 4;
	else return 0;
}

int main() {
	cin.tie(0);
	ios_base::sync_with_stdio(false);

	stack<char> s;
	string input;
	cin >> input;
	input = '(' + input + ')';
	for (auto i : input) {
		if (i == '(') s.push(i);
		else if (p(i) > 0) {
			while (!s.empty() && p(s.top()) <= p(i)) {
				cout << s.top();
				s.pop();
			}
			if (i == ')') s.pop();
			else s.push(i);
		}
		else cout << i;
	}
}

File: BOJ/1918.py
# boj1918 후위 표기식 풀이
# 1. 여는 괄호: 스택에 넣는다.
# 2. 닫는 괄호: 여는 괄호가 나올 때까지 스택에서 연산자를 pop하여 후위 표기식에 추가한다.
# -> 괄호 안에 있는 연산은 가장 먼저 실행하기 때문

# 3. 연산자:
#       a. 현재 연산자보다 우선순위가 낮은 연산자가 올 때까지 스택에서 연산자를 pop하여 후위 표기식에 추가한다.
#       b. 스택에 현재 연산자를 넣는다.
# -> */의 우선순위가 +-의 우선순위보다 낮다.

# 4. 피연산자: 바로 후위 표기식에 추가한다.

priority = {'*': 1, '/': 1, # 우선순위가 높음
            '+': 2, '-': 2} # 우선순위가 낮음

def get_postfix(infix):
    stack = []
    ret = ''
    for c in infix:
        # 1. 여는 괄호
        if c == '(':
            stack.append(c)
        # 2. 닫는 괄호
        elif c == ')':
            while stack[-1] != '(':
                ret += stack.pop()
            stack.pop()
        # 3. 연산자
        elif c in priority:
            while stack and stack[-1] in priority and priority[stack[-1]] <= priority[c]:
                ret += stack.pop()
            else:
                stack.append(c)
        # 4. 피연산자
        else:
            ret += c

    # 남은 연산자들을 후위 표기식에 추가
    while stack:
        ret += stack.pop()

    return ret


if __name__ == '__main__':
    infix = input()
    postfix = get_postfix(infix)
    print(postfix)